PEP 612: Parameter Specification Variables

mrkmndz opened the pull request:
This is a pull request for a new PEP proposal for adding a new category of type variables to the PEP 484 typing ecosystem -- ParameterSpecifications.

Some procedural notes:

1. Although this PEP was not discussed in python-ideas, it was extensively discussed by the attendees of the last Python typing summit, where it reached general consensus, on everything except for the name for the feature itself, where there was some disagreement.  I anticipate some discussion on this front in typing-sig.
2. This has been implemented into Pyre, missing only handling for user-defined generic classes.
3. I discussed this submission ahead of time with @ilevkivskyi, a core contributor, could I add you as a sponsor?
4. It looks like 612 is the next free number, but I may not be looking in all of the relevant places.

sobolevn commented on pep-0612.rst:
```diff
@@ -0,0 +1,258 @@
+Mutations on ParameterSpecifications
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There are still a class of decorators still not supported with these features: those that mutate (add/remove/change) the parameters of the given function. Defining operators that do these mutations becomes very complicated very quickly, as you have to deal with name collision issues much more prominently. We will defer that work until there is significant demand, and then we would be open to revi ...
```
I would say that this is needed, there is even a whole family of decorators that do just that. It seems like `ParameterSpecification` does not leave an open space for this feature in the future. And can even be conflicting with it.

Real world examples of decorators that do change both input and output types: https://github.com/dry-python/returns/blob/master/returns/contrib/mypy/decorator_plugin.py

mrkmndz replied:
Would you mind explaining this example in a bit more detail?  To me it appears that those decorators are propagating the entire set of parameters over, without any mutation. 

By mutation of parameters, I was referring to something like this: (removal)
```
def decorator(f):
    def inner(*args, **kwargs):
         return f(extra=1, *args, **kwargs)
    return inner
```
or this (addition)
```
def decorator(f):
    def inner(extra, *args, **kwargs):
         return f(*args, **kwargs)
    return inner
```

sobolevn replied:
In this particular case we often change both input and return types: `T` -> `Wrapper[T]`, where wrapper can be `List`, `Optional`, `Maybe`, `IO`, etc.

mrkmndz replied:
Are you referring to a decorator which turns a `Callable[[int, str], bool]` into a `Callable[[Wrapper[int], Wrapper[str]], Wrapper[bool]]`?  
Would you mind giving a me a snippet or pointer to an instance of this so I can better understand what you're describing here?

sobolevn replied:
Sure: 

- https://returns.readthedocs.io/en/latest/pages/maybe.html#maybe-decorator
- https://returns.readthedocs.io/en/latest/pages/io.html#impure
- https://returns.readthedocs.io/en/latest/pages/result.html#safe

mrkmndz replied:
All three of those appear to be only modifying the return types of the decorated functions.
This means, that with ParameterSpecifications, you should be able to type them as 
```
TParams = typing.ParameterSpecification("TParams")
def maybe(
    function: Callable[TParams, Optional[_ValueType]],
) -> Callable[TParams, Maybe[_ValueType]]: ...
```

```
def impure(
    function: Callable[TParams, _NewValueType],
) -> Callable[TParams, IO[_NewValueType]]:....
```

```
def safe(
    function: Callable[TParams, _ValueType],
) -> Callable[TParams, Result[_ValueType, Exception]]: ...
```

with overloads for coroutines like
```
def maybe(
    function: Callable[
        TParams,
        Coroutine[_FirstType, _SecondType, Optional[_ValueType]],
    ],
) -> Callable[
    TParams,
    Coroutine[_FirstType, _SecondType, Maybe[_ValueType]],
]:
```
and so on for the other ones.

Will that work for your use cases?

sobolevn replied:
Fair enough! This would work for my cases. 

Thanks a lot for taking your time and explaining this. 👍

mrkmndz replied:
No problem!  Thanks for taking a look at the PEP.  As @Rosuav said, if you have any other questions/comments/concerns about this, we should discuss them on typing-sig@python.org (https://mail.python.org/archives/list/typing-sig@python.org/thread/UDHSH4EVVIDKNLRX2YGCIUCBGZ5ALRKC/)

asvetlov commented:
I still think that modifying the parameters list signature is important.
One very common example is `unittest.mock`:
```
@patch('__main__.SomeClass')
def function(normal_argument, mock_class):
    print(mock_class is SomeClass)
```
Another one is `click`:
```
@cli.command()
@click.pass_context
def sync(ctx):
    click.echo('Debug is %s' % (ctx.obj['DEBUG'] and 'on' or 'off'))
```
Note, `pass_context` decorator adds a parameter to the function signature. I can imagine the reverse operation easily.

asvetlov commented:
Currently, Python supports 5 types of arguments: 
* positional-only
* regular arguments that can be used in both positional and named contexts
* keyword-only
* var-args: `*args`
* keyword var-args: `*kwargs`

The first three can have default arguments, it also affects a signature.
I think it makes the proposed object much more complex

Rosuav commented:
Can we get the PEP merged into the repository before going into too much detail about the proposal, please? Let's keep comments on this PR to matters that would stop the merge.

Now that the formatting issues are sorted out, are there any remaining problems? If not, let's just merge in and then let the discussion happen on typing-sig (as per the PEP header).

mrkmndz commented:
@Rosuav That sounds like it would be best, is there anything else we need to do before merging?

@asvetlov on that note, would you mind sending your message to typing-sig so we don't get wires crossed?

Rosuav commented:
@mrkmndz One possible consideration: You're using the old wording for the copyright statement, but PEP 12 now recommends slightly different wording. I don't know who all the authors/contributors are, but perhaps you know who else has made substantive textual contributions. Are you okay with the newer wording (which mentions CC0), and can you tag all other people who would hold copyright on any portion of this text, asking if they also are okay with the change?

sk- commented on pep-0612.rst:
```diff
@@ -0,0 +1,258 @@
+
+With these three properties, we now have the ability to fully type check parameter preserving decorators.
+
+One additional form that we want to support is functions that pass only a subset of their arguments on to another function. To avoid shadowing a named or keyword only argument in the ``ParameterSpecification`` we require that the additional arguments be anonymous arguments that precede the ``*args`` and ``*kwargs``
```
What do you mean by `anonymous` arguments? Are those arguments prefixed by double underscore (`__`) as in the provided example. Is that rule enforced or suggested?

Rosuav replied:
This PR has been merged. For discussions about the substance of the proposal, please post to typing-sig.

(If you were half way through writing this when the merge went through, then my apologies; you'll want to repost to typing-sig to get responses.)

mrkmndz commented:
I think I see the better license wording you're talking about, and since I'm the only copyright holder (I got feedback from others, but did the wordsmithing) I've just amended it.

sk- commented on pep-0612.rst:
```diff
@@ -0,0 +1,258 @@
+The components of a ``ParameterSpecification``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A ``ParameterSpecification`` captures both positional and keyword accessible parameters, but there unfortunately is no object in the runtime that captures both of these together. Instead, we are forced to separate them into ``*args`` and ``**kwargs``\ , respectively. This means we need to be able to split apart a single ``ParameterSpecification`` into these two components, and then bring them bac ...
```
What is the motivation to require `TParams.args` and `TParams.kwargs` to be used together.

Why is the example `def baz(*args: TParams.args) -> int` rejected?

Rosuav commented:
LGTM! Further discussion regarding content, wording, whether or not this should happen, etc, can happen on typing-sig. The PEP should go live shortly.

mrkmndz commented:
Thanks @Rosuav!

@asvetlov I responded to your two comments above in the typing-sig list [here](https://mail.python.org/archives/list/typing-sig@python.org/message/N4ESM3Y2H3UURL3V2KGORMR3FBMVMSLQ/), please let me know if that addresses your concerns

sk- commented on pep-0612.rst:
```diff
@@ -0,0 +1,258 @@
+
+Just as with normal ``TypeVar``\ s, ``ParameterSpecification``\ s can be used to make generic classes as well as generic functions. These ought to be able to be mixed with normal ``TypeVar``\ s. This should also be made to work with protocols in the same manner.
+
+The components of a ``ParameterSpecification``
```
I'm missing a brief description on what happens to keyword only ([PEP 3102](https://www.python.org/dev/peps/pep-3102/)) and to positional only ([PEP 570](https://www.python.org/dev/peps/pep-0570/)) arguments.

Will for example the following be accepted or rejected:
```
def add_logging(f: Callable[Ps, R]) -> Callable[Ps, R]:
       async def inner(*args: Ps.args, **kwargs: Ps.kwargs) -> R:
           # log
           return f(*args, **kwargs)
       return inner

@add_logging
def add(x:int, *, y: int = 0):
  return x + y

add(1, 2)
```
Note that today in Mypy, that `add` call (without the decorator) will be rejected.

I only found there's some talk abut it in the rejected proposals section.

gvanrossum commented on pep-0612.rst:
```diff
@@ -0,0 +1,258 @@
+
+   from typing import Awaitable, Callable, ParameterSpecification, TypeVar
+
+   Ps = ParameterSpecification("TParams")
```
I think this should have `"Ps"` as the call argument, not `"TParams"`.

mrkmndz replied:
Corrected in https://github.com/python/peps/pull/1260

gvanrossum commented on pep-0612.rst:
```diff
@@ -0,0 +1,258 @@
+       return x + 7
+
+   await foo(1, "A")
+   await foo("B", 2) # Incompatible parameter type: Expected `int` for 1st anonymous parameter 
```
This comment is too long and causes ugly folding in the rendered PEP. There are a few others that also overflow; you can see the rendered PEP at https://www.python.org/dev/peps/pep-0612/ to find them.

mrkmndz replied:
re-flow-ed all of the comments in https://github.com/python/peps/pull/1260

gvanrossum commented on pep-0612.rst:
```diff
@@ -0,0 +1,258 @@
+Abstract
+--------
+
+There currently are two ways to specify the type of a callable, the ``Callable[[T1, T2], TReturn]`` syntax defined in  `PEP 484 <https://www.python.org/dev/peps/pep-0484>`_\ , and callback protocols from `PEP 544 <https://www.python.org/dev/peps/pep-0544/#callback-protocols>`_. Neither of these support forwarding the parameter types of one callable over to another callable, making it difficult to ...
```
I honestly prefer if the raw text of a PEP is wrapped at something like 70-80 columns (with an exception for long URLs etc.).

PS. Why the backslash-space before commas?

mrkmndz replied:
re: the escaping, I'm not exactly sure, as they were generated by this markdown -> rst converter (https://pypi.org/project/m2r/), but removing them causes this error
```
 (WARNING/2) Inline literal start-string without end-string.
```

mrkmndz replied:
I wrapped at 80 in https://github.com/python/peps/pull/1260.

sk- commented:
I also added some questions about the PEP. See
https://github.com/python/peps/pull/1259#pullrequestreview-335307738.

On Fri, Dec 20, 2019 at 1:54 PM Guido van Rossum <notifications@github.com>
wrote:

> *@gvanrossum* commented on this pull request.
>
> I have a few *formatting* comments that you may address in a separate PR.
>
> I agree that the *discussion* of the proposal ought to go to typing-sig.
> ------------------------------
>
> In pep-0612.rst
> <https://github.com/python/peps/pull/1259#discussion_r360464623>:
>
> > +   await foo(1, "A")
>
> +   await foo("B", 2) # fails at runtime
>
> +
>
> +``add_logging``\ , a decorator which logs before each entry into the decorated function, is an instance of the Python idiom of one function passing all arguments given to it over to another function through the combination of the ``*args`` and ``**kwargs`` features in both parameters and in arguments. When one defines a function (like ``inner``\ ) that takes ``(*args, **kwargs)`` and goes on to call another function with ``(*args, **kwargs)``\ , the wrapping function can only be safely called in all of the ways that the wrapped function could be safely called. To type this decorator, we’d like to be able to place a dependency between the parameters of the callable ``f`` and the parameters of the returned function. `PEP 484 <https://www.python.org/dev/peps/pep-0484>`_  supports dependencies between single types, as in ``def append(l: typing.List[T], e: T) -> typing.List[T]: ...``\ , but there is no existing way to do so with a complicated entity like the parameters one could pass to a function.
>
> +
>
> +Due to the limitations of the status quo, the ``add_logging`` example will type check but will fail at runtime. ``inner`` will pass the string “B” into ``foo``\ , which will try to add 7 to it, triggering a type error.  This was not caught by the type checker because the decorated ``foo`` was given the type ``Callable[..., Awaitable[int]]`` which is specified to do no validation on its arguments.
>
> +
>
> +Without the ability to define dependencies between the parameters of different callable types, there is no way, at present, to make ``add_logging`` compatible with all functions, while still preserving the enforcement of the parameters of the decorated function.
>
> +
>
> +With the addition of the ``ParameterSpecification`` variables proposed by this PEP, we can rewrite the previous example in a way that keeps the flexibility of the decorator and the parameter enforcement of the decorated function.
>
> +
>
> +.. code-block::
>
> +
>
> +   from typing import Awaitable, Callable, ParameterSpecification, TypeVar
>
> +
>
> +   Ps = ParameterSpecification("TParams")
>
>
> I think this should have "Ps" as the call argument, not "TParams".
> ------------------------------
>
> In pep-0612.rst
> <https://github.com/python/peps/pull/1259#discussion_r360465758>:
>
> > +   Ps = ParameterSpecification("TParams")
>
> +   R = TypeVar("TReturn")
>
> +
>
> +   def add_logging(f: Callable[Ps, R]) -> Callable[Ps, Awaitable[R]]:
>
> +       async def inner(*args: Ps.args, **kwargs: Ps.kwargs) -> R:
>
> +           await log_to_database()
>
> +           return f(*args, **kwargs)
>
> +
>
> +       return inner
>
> +
>
> +   @add_logging
>
> +   def foo(x: int, y: str) -> int:
>
> +       return x + 7
>
> +
>
> +   await foo(1, "A")
>
> +   await foo("B", 2) # Incompatible parameter type: Expected `int` for 1st anonymous parameter
>
>
> This comment is too long and causes ugly folding in the rendered PEP.
> There are a few others that also overflow; you can see the rendered PEP at
> https://www.python.org/dev/peps/pep-0612/ to find them.
> ------------------------------
>
> In pep-0612.rst
> <https://github.com/python/peps/pull/1259#discussion_r360467395>:
>
> > +Author: Mark Mendoza <mendoza.mark.a@gmail.com>
>
> +Discussions-To: Typing-Sig <typing-sig@python.org>
>
> +Status: Draft
>
> +Type: Standards Track
>
> +Content-Type: text/x-rst
>
> +Created: 18-Dec-2019
>
> +Python-Version: 3.9
>
> +Post-History: 18-Dec-2019
>
> +
>
> +Parameter Specification Variables
>
> +=================================
>
> +
>
> +Abstract
>
> +--------
>
> +
>
> +There currently are two ways to specify the type of a callable, the ``Callable[[T1, T2], TReturn]`` syntax defined in  `PEP 484 <https://www.python.org/dev/peps/pep-0484>`_\ , and callback protocols from `PEP 544 <https://www.python.org/dev/peps/pep-0544/#callback-protocols>`_. Neither of these support forwarding the parameter types of one callable over to another callable, making it difficult to annotate function decorators. This PEP proposes ``typing.ParameterSpecification``\ , a new kind of type variable, to support expressing these kinds of relationships.
>
>
> I honestly prefer if the raw text of a PEP is wrapped at something like
> 70-80 columns (with an exception for long URLs etc.).
>
> PS. Why the backslash-space before commas?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/peps/pull/1259?email_source=notifications&email_token=AAG6TGFKG2CXJJNPVHPGDRLQZT2F5A5CNFSM4J5R6PQKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCP6XPHY#pullrequestreview-335378335>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG6TGDVDK72CAD27GACJVDQZT2F5ANCNFSM4J5R6PQA>
> .
>


-- 
Sebastian Kreft

mrkmndz commented:
@sk- would you mind transferring your questions over to the thread on the typing-sig list?
https://mail.python.org/archives/list/typing-sig@python.org/thread/UDHSH4EVVIDKNLRX2YGCIUCBGZ5ALRKC/

sk- commented:
I do have some questions regarding the proposed PEP. Those questions were
being written while the PR was merged.

1)

I'm missing a brief description on what happens to keyword only (PEP 3102
<https://www.python.org/dev/peps/pep-3102/>) and to positional only (PEP 570
<https://www.python.org/dev/peps/pep-0570/>) arguments.

Will for example the following be accepted or rejected:

def add_logging(f: Callable[Ps, R]) -> Callable[Ps, R]:
       async def inner(*args: Ps.args, **kwargs: Ps.kwargs) -> R:
           # log
           return f(*args, **kwargs)
       return inner

@add_logging
def add(x:int, *, y: int = 0):
  return x + y

add(1, 2)

Note that today in Mypy, that add call (without the decorator) will be
rejected.

I only found there's some talk abut it in the rejected proposals section.

2) What do you mean by anonymous arguments? Are those arguments prefixed by
double underscore (__) as in the provided example. Is that rule enforced or
suggested?

3) What is the motivation to require TParams.args and TParams.kwargs to be
used together.

Why is the example def baz(*args: TParams.args) -> int rejected?

On Fri, Dec 20, 2019 at 3:25 PM Sebastian Kreft <skreft@gmail.com> wrote:

> I also added some questions about the PEP. See
> https://github.com/python/peps/pull/1259#pullrequestreview-335307738.
>
> On Fri, Dec 20, 2019 at 1:54 PM Guido van Rossum <notifications@github.com>
> wrote:
>
>> *@gvanrossum* commented on this pull request.
>>
>> I have a few *formatting* comments that you may address in a separate PR.
>>
>> I agree that the *discussion* of the proposal ought to go to typing-sig.
>> ------------------------------
>>
>> In pep-0612.rst
>> <https://github.com/python/peps/pull/1259#discussion_r360464623>:
>>
>> > +   await foo(1, "A")
>>
>> +   await foo("B", 2) # fails at runtime
>>
>> +
>>
>> +``add_logging``\ , a decorator which logs before each entry into the decorated function, is an instance of the Python idiom of one function passing all arguments given to it over to another function through the combination of the ``*args`` and ``**kwargs`` features in both parameters and in arguments. When one defines a function (like ``inner``\ ) that takes ``(*args, **kwargs)`` and goes on to call another function with ``(*args, **kwargs)``\ , the wrapping function can only be safely called in all of the ways that the wrapped function could be safely called. To type this decorator, we’d like to be able to place a dependency between the parameters of the callable ``f`` and the parameters of the returned function. `PEP 484 <https://www.python.org/dev/peps/pep-0484>`_  supports dependencies between single types, as in ``def append(l: typing.List[T], e: T) -> typing.List[T]: ...``\ , but there is no existing way to do so with a complicated entity like the parameters one could pass to a function.
>>
>> +
>>
>> +Due to the limitations of the status quo, the ``add_logging`` example will type check but will fail at runtime. ``inner`` will pass the string “B” into ``foo``\ , which will try to add 7 to it, triggering a type error.  This was not caught by the type checker because the decorated ``foo`` was given the type ``Callable[..., Awaitable[int]]`` which is specified to do no validation on its arguments.
>>
>> +
>>
>> +Without the ability to define dependencies between the parameters of different callable types, there is no way, at present, to make ``add_logging`` compatible with all functions, while still preserving the enforcement of the parameters of the decorated function.
>>
>> +
>>
>> +With the addition of the ``ParameterSpecification`` variables proposed by this PEP, we can rewrite the previous example in a way that keeps the flexibility of the decorator and the parameter enforcement of the decorated function.
>>
>> +
>>
>> +.. code-block::
>>
>> +
>>
>> +   from typing import Awaitable, Callable, ParameterSpecification, TypeVar
>>
>> +
>>
>> +   Ps = ParameterSpecification("TParams")
>>
>>
>> I think this should have "Ps" as the call argument, not "TParams".
>> ------------------------------
>>
>> In pep-0612.rst
>> <https://github.com/python/peps/pull/1259#discussion_r360465758>:
>>
>> > +   Ps = ParameterSpecification("TParams")
>>
>> +   R = TypeVar("TReturn")
>>
>> +
>>
>> +   def add_logging(f: Callable[Ps, R]) -> Callable[Ps, Awaitable[R]]:
>>
>> +       async def inner(*args: Ps.args, **kwargs: Ps.kwargs) -> R:
>>
>> +           await log_to_database()
>>
>> +           return f(*args, **kwargs)
>>
>> +
>>
>> +       return inner
>>
>> +
>>
>> +   @add_logging
>>
>> +   def foo(x: int, y: str) -> int:
>>
>> +       return x + 7
>>
>> +
>>
>> +   await foo(1, "A")
>>
>> +   await foo("B", 2) # Incompatible parameter type: Expected `int` for 1st anonymous parameter
>>
>>
>> This comment is too long and causes ugly folding in the rendered PEP.
>> There are a few others that also overflow; you can see the rendered PEP at
>> https://www.python.org/dev/peps/pep-0612/ to find them.
>> ------------------------------
>>
>> In pep-0612.rst
>> <https://github.com/python/peps/pull/1259#discussion_r360467395>:
>>
>> > +Author: Mark Mendoza <mendoza.mark.a@gmail.com>
>>
>> +Discussions-To: Typing-Sig <typing-sig@python.org>
>>
>> +Status: Draft
>>
>> +Type: Standards Track
>>
>> +Content-Type: text/x-rst
>>
>> +Created: 18-Dec-2019
>>
>> +Python-Version: 3.9
>>
>> +Post-History: 18-Dec-2019
>>
>> +
>>
>> +Parameter Specification Variables
>>
>> +=================================
>>
>> +
>>
>> +Abstract
>>
>> +--------
>>
>> +
>>
>> +There currently are two ways to specify the type of a callable, the ``Callable[[T1, T2], TReturn]`` syntax defined in  `PEP 484 <https://www.python.org/dev/peps/pep-0484>`_\ , and callback protocols from `PEP 544 <https://www.python.org/dev/peps/pep-0544/#callback-protocols>`_. Neither of these support forwarding the parameter types of one callable over to another callable, making it difficult to annotate function decorators. This PEP proposes ``typing.ParameterSpecification``\ , a new kind of type variable, to support expressing these kinds of relationships.
>>
>>
>> I honestly prefer if the raw text of a PEP is wrapped at something like
>> 70-80 columns (with an exception for long URLs etc.).
>>
>> PS. Why the backslash-space before commas?
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/python/peps/pull/1259?email_source=notifications&email_token=AAG6TGFKG2CXJJNPVHPGDRLQZT2F5A5CNFSM4J5R6PQKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCP6XPHY#pullrequestreview-335378335>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAG6TGDVDK72CAD27GACJVDQZT2F5ANCNFSM4J5R6PQA>
>> .
>>
>
>
> --
> Sebastian Kreft
>


-- 
Sebastian Kreft

mrkmndz commented:
Hi @sk- , I responded to your questions on the typing-sig thread [here](https://mail.python.org/archives/list/typing-sig@python.org/message/HDSNIE3BPKQMJMHYRRMZDYLZLZI7HTAM/), please let me know if that addresses your concerns.

mrkmndz commented:
@gvanrossum Thanks for the formatting review!  I think I got everything into https://github.com/python/peps/pull/1260

gvanrossum commented:
Indeed. Thanks!

On Fri, Dec 20, 2019 at 13:45 Mark Mendoza <notifications@github.com> wrote:

> @gvanrossum <https://github.com/gvanrossum> Thanks for the formatting
> review! I think I got everything into #1260
> <https://github.com/python/peps/pull/1260>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/python/peps/pull/1259?email_source=notifications&email_token=AAWCWMQIS34PADH6UO2YZATQZU4GBA5CNFSM4J5R6PQKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHOIDIA#issuecomment-568099232>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWCWMUWFT7OYHKMCV2QYS3QZU4GBANCNFSM4J5R6PQA>
> .
>
-- 
--Guido (mobile)
